src/pycolmap/sfm/incremental_mapper.cc

```diff
@@ -96,3 +158,146 @@ void BindIncrementalMapper(py::module& m) {
+                 ImageSelection::MAX_VISIBLE_POINTS_NUM)
+          .value("MAX_VISIBLE_POINTS_RATIO",
+                 ImageSelection::MAX_VISIBLE_POINTS_RATIO)
+          .value("MIN_UNCERTAINTY", ImageSelection::MIN_UNCERTAINTY);
+  AddStringToEnumConstructor(PyImageSelectionMethod);
+
+  // bind local bundle adjustment report
+  using LocalBAReport = IncrementalMapper::LocalBundleAdjustmentReport;
+  auto PyLocalBAReport =
+      py::class_<LocalBAReport>(m, "LocalBundleAdjustmentReport");
+  PyLocalBAReport.def(py::init<>())
+      .def_readwrite("num_merged_observations",
+                     &LocalBAReport::num_merged_observations)
+      .def_readwrite("num_completed_observations",
+                     &LocalBAReport::num_completed_observations)
+      .def_readwrite("num_filtered_observations",
+                     &LocalBAReport::num_filtered_observations)
+      .def_readwrite("num_adjusted_observations",
+                     &LocalBAReport::num_adjusted_observations);
+  MakeDataclass(PyLocalBAReport);
+
+  // bind incremental mapper
+  // TODO: migrate comments. improve formatting
+  py::class_<IncrementalMapper, std::shared_ptr<IncrementalMapper>>(
+      m, "IncrementalMapper")
+      .def(py::init<std::shared_ptr<const DatabaseCache>>())
+      .def("begin_reconstruction",
+           &IncrementalMapper::BeginReconstruction,
+           "reconstruction"_a)
+      .def("end_reconstruction",
+           &IncrementalMapper::EndReconstruction,
+           "discard"_a)
+      .def(
+          "find_initial_image_pair",  // explicitly handle the conversion from
+                                      // -1 (int) to image_t (uint32_t)
+          [](IncrementalMapper& self,
+             const IncrementalMapper::Options& options,
+             int image_id1,
+             int image_id2) {
+            TwoViewGeometry two_view_geometry;
+            image_t image_id1_cast = image_id1;
+            image_t image_id2_cast = image_id2;
+            bool success = self.FindInitialImagePair(
+                options, two_view_geometry, image_id1_cast, image_id2_cast);
+            return std::make_pair(
+                success,
+                std::make_pair(std::make_pair(image_id1_cast, image_id2_cast),
+                               two_view_geometry));
+          },
+          "options"_a,
+          "image_id1"_a,
+          "image_id2"_a)
+      .def(
+          "estimate_initial_two_view_geometry",
+          [](IncrementalMapper& self,
+             const IncrementalMapper::Options& options,
+             const image_t image_id1,
+             const image_t image_id2) {
+            TwoViewGeometry two_view_geometry;
+            bool success = self.EstimateInitialTwoViewGeometry(
+                options, two_view_geometry, image_id1, image_id2);
+            if (success)
+              return py::cast(two_view_geometry);
+            else
+              return py::cast<py::object>(Py_None);
```

Comment:
Use `py::none()` instead?

https://github.com/colmap/colmap/blob/06a805386afb3db9dc6704c7ba2a80861bb64f7e/src/pycolmap/estimators/absolute_pose.cc#L38

Reply:
Lambda function needs a consistent type of output. I was using ``py::none()`` before but got this unfortunately:

error: inconsistent types ‘pybind11::object’ and ‘pybind11::none’ deduced for lambda return type.

So I changed into the current design.

Reply:
How about if you annotate the return type with `-> py::object`? Then you maybe don’t even need the other `py::cast`

Reply:
Nice idea. This is done. the ``py::cast`` on TwoViewGeometry object is still needed though.